src/libproc_macro/lib.rs

```diff
@@ -191,52 +181,34 @@ impl Extend<TokenTree> for TokenStream {
 #[stable(feature = "token_stream_extend", since = "1.30.0")]
 impl Extend<TokenStream> for TokenStream {
     fn extend<I: IntoIterator<Item = TokenStream>>(&mut self, streams: I) {
-        self.0.extend(streams.into_iter().map(|stream| stream.0));
+        // FIXME(eddyb) Use an optimized implementation if/when possible.
+        *self = iter::once(mem::replace(self, Self::new())).chain(streams).collect();
```

Comment:
I can't think of a way other than creating a `Vec<TokenStream>` (which should now be possible to pass to the server with the new serialization approach).
Would that work for you? Are there benchmarks for this?

Reply:
My real-world benchmark for proc macro changes has been:

```console
$ git clone https://github.com/serde-rs/serde
$ cd serde/test_suite
$ cargo check --tests --features proc-macro2/nightly
$ rm -f ../target/debug/deps/libtest_*.rmeta
$ time cargo check --tests --features proc-macro2/nightly
```

Reply:
> Could a new API bridge method be added for something like 

Sure, but isn't that what `fn push($self: &mut $S::TokenStreamBuilder, stream: $S::TokenStream);` does?
And that's what I'm using here by doing `iter::once(self).chain(streams).collect()`.
That is, it will end up calling `TokenStreamBuilder::push` a total of `streams.count() + 1` times, irrespective of how many tokens each token stream has.

Or are you saying that we should bypass the builder altogether here, because that would be more efficient?

@dtolnay I think we should get https://github.com/rust-lang/rust/pull/49219#issuecomment-414181092 fixed and use perf.r-l.o for this, wdyt @Mark-Simulacrum @kennytm?

Reply:
Should be feasible; happy to take a PR or an issue on rustc-perf.